presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -223,11 +224,22 @@ public PlanWithProperties visitOutput(OutputNode node, Context context)
         {
             PlanWithProperties child = planChild(node, context.withPreferredProperties(PreferredProperties.undistributed()));
 
+            // if the child is already a gathering exchange, don't add another exchange
+            if ((child.getNode() instanceof ExchangeNode) && ((ExchangeNode) child.getNode()).getType() == ExchangeNode.Type.GATHER) {
```

Comment:
`GATHER` exchange does not necessarily mean `child.getProperties().isCoordinatorOnly()==true`